jwst/ramp_fitting/ramp_fit.py

```diff
@@ -538,72 +527,11 @@ def ols_ramp_fit(data, err, groupdq, inpixeldq, buffsize, save_opt, readnoise_2d
-
-        while (np.all(np.bitwise_and( first_gdq, dqflags.group['DO_NOT_USE']))):
-            num_bad_slices += 1
-            nreads -= 1
-            ngroups -= 1
-
-            # Check if there are remaining groups before accessing data
-            if ngroups < 1 :  # no usable data
-                log.error('1. All groups have all pixels flagged as DO_NOT_USE,')
-                log.error('  so will not process this dataset.')
-                return None, None, None
-
-            data = data[:,1:,:,:]
-            err = err[:,1:,:,:]
-            groupdq = groupdq[:,1:,:,:]
-
-            cubeshape = (nreads,) + imshape
-
-            # Where the initial group of the just-truncated data is a cosmic ray,
-            #   remove the JUMP_DET flag from the group dq for those pixels so
-            #   that those groups will be included in the fit.
-            wh_cr = np.where( np.bitwise_and(groupdq[:,0,:,:],
-                              dqflags.group['JUMP_DET']) != 0 )
-            num_cr_1st = len(wh_cr[0])
-
-            for ii in range(num_cr_1st):
-                groupdq[ wh_cr[0][ii], 0, wh_cr[1][ii],
-                    wh_cr[2][ii]] -=  dqflags.group['JUMP_DET']
-
-            first_gdq = groupdq[:,0,:,:]
-
-        log.info('Number of leading groups that are flagged as DO_NOT_USE: %s', num_bad_slices)
-
-        # If all groups were flagged, the final group would have been picked up
-        #   in the while loop above, ngroups would have been set to 0, and Nones
-        #   would have been returned.  If execution has gotten here, there must
-        #   be at least 1 remaining group that is not all flagged.
-        last_gdq = groupdq[:,-1,:,:]
-        if np.all(np.bitwise_and( last_gdq, dqflags.group['DO_NOT_USE'] )):
-            nreads -= 1
-            ngroups -= 1
-
-            # Check if there are remaining groups before accessing data
-            if ngroups < 1 :  # no usable data
-                log.error('2. All groups have all pixels flagged as DO_NOT_USE,')
-                log.error('  so will not process this dataset.')
-                return None, None, None
-
-            data = data[:,:-1,:,:]
-            err = err[:,:-1,:,:]
-            groupdq = groupdq[:,:-1,:,:]
-            cubeshape = (nreads,)+imshape
-
-            log.info('MIRI dataset has all pixels in the final group flagged as DO_NOT_USE.')
-
-        # Next block is to satisfy github issue 1681:
-        # "MIRI FirstFrame and LastFrame minimum number of groups"
-        if (ngroups < 2):
-            log.warning('MIRI datasets require at least 2 groups/integration')
-            log.warning('(NGROUPS), so will not process this dataset.')
-            return None, None, None
+        miri_ans = ramp_fit_miri_multigroups(data, err, groupdq, cubeshape, imshape, nreads, ngroups)
+        if miri_ans is None:
+            return None * 22
+        data, err, groupdq, cubeshape, imshape, nreads, ngroups, first_gdq = miri_ans
```

Comment:
`miri_ans` could be a little clearer what it is, though this can maybe be cleaned up when there's a data class for the returned items.

Is this `ramp_fit_miri_multigroups()` function only specific to MIRI?  If not, it might be good to give it a more general name, especially if it might be useful for Roman or any other sort of ramp fitting.

Reply:
This function is only called for MIRI images that have more than 1 nreads.  The shape of the 4-D cube gets changed based on removing beginning and end groups in MIRI images.